internal/agent/device/spec/spec_test.go

```diff
@@ -551,13 +553,525 @@ func TestCheckOsReconciliation(t *testing.T) {
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, deviceName, params).Return(respSpec, http.StatusOK, nil)
+
+		rendered := &v1alpha1.RenderedDeviceSpec{}
+		success, err := s.getRenderedFromManagementAPIWithRetry(ctx, "", rendered)
+		require.NoError(err)
+		require.True(success)
+		require.Equal(respSpec, rendered)
+	})
+
+	t.Run("makes a request with the passed renderedVersion when set", func(tt *testing.T) {
+		respSpec := createRenderedTestSpec("requested-image:latest")
+		renderedVersion := "24"
+		params := &v1alpha1.GetRenderedDeviceSpecParams{KnownRenderedVersion: &renderedVersion}
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, deviceName, params).Return(respSpec, http.StatusOK, nil)
+
+		rendered := &v1alpha1.RenderedDeviceSpec{}
+		success, err := s.getRenderedFromManagementAPIWithRetry(ctx, "24", rendered)
+		require.NoError(err)
+		require.True(success)
+		require.Equal(respSpec, rendered)
+	})
+}
+
+func Test_pathFromType(t *testing.T) {
+	require := require.New(t)
+
+	s := &SpecManager{
+		currentPath:  "test/current.json",
+		desiredPath:  "test/desired.json",
+		rollbackPath: "test/rollback.json",
+	}
+
+	testCases := []struct {
+		Name          string
+		SpecType      Type
+		ExpectedPath  string
+		ExpectedError error
+	}{
+		{"current", "current", s.currentPath, nil},
+		{"desired", "desired", s.desiredPath, nil},
+		{"rollback", "rollback", s.rollbackPath, nil},
+		{"invalid spec type", "rainbow", "", ErrInvalidSpecType},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			path, err := s.pathFromType(testCase.SpecType)
+
+			if testCase.ExpectedError != nil {
+				require.ErrorIs(err, testCase.ExpectedError)
+			} else {
+				require.NoError(err)
+			}
+
+			require.Equal(testCase.ExpectedPath, path)
+		})
+	}
+}
+
+func Test_getNextRenderedVersion(t *testing.T) {
+	require := require.New(t)
+	testCases := []struct {
+		Name                string
+		RenderedVersion     string
+		NextRenderedVersion string
+		ExpectedError       error
+	}{
+		{"empty rendered version", "", "", nil},
+		{"increments the rendered version", "1", "2", nil},
+		{"errors when the rendered version cannot be parsed", "not-a-number", "", ErrParseRenderedVersion},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			nextVersion, err := getNextRenderedVersion(testCase.RenderedVersion)
+
+			if testCase.ExpectedError != nil {
+				require.ErrorIs(err, testCase.ExpectedError)
+			} else {
+				require.NoError(err)
+			}
+
+			require.Equal(testCase.NextRenderedVersion, nextVersion)
+		})
+	}
+}
+
+func Test_getRenderedVersion(t *testing.T) {
+	require := require.New(t)
+
+	s := &SpecManager{
+		log: log.NewPrefixLogger("test"),
+	}
+
+	testCases := []struct {
+		Name                    string
+		CurrentRenderedVersion  string
+		DesiredRenderedVersion  string
+		RollbackRenderedVersion string
+		ExpectedReturnValue     string
+		ExpectedError           error
```

Comment:
minor:

Generally, this test structure (which I like) uses non-exported keys and for readability use the key name in the case. if expectedError is nil you can just leave it out. `name` in this case is the name of the test so adding a bit more context is helpful when you are trying to debug.

```
func Test_pathFromType(t *testing.T) {
	require := require.New(t)

	s := &SpecManager{
		currentPath:  "test/current.json",
		desiredPath:  "test/desired.json",
		rollbackPath: "test/rollback.json",
	}

	testCases := []struct {
		name          string
		specType      Type
		expectedPath  string
		expectedError error
	}{
		{
			name:"current",
			specType:"current",
			expectedPath:s.currentPath,
		},
		...
	```